### ISSUE TITLE
Discontinuous FEM shape functions in MeshFunction

### ISSUE THREAD

**Issue by snschune** [libMesh/libmesh#1011]
Related to idaholab/moose#7244

When discontinuous FEM solutions are evaluated on the face, the value is not unique. Therefore,
MeshFunction should have a method that returns all values residing in elements that are close to
this point. Note: if you are close to a mesh node, there might be more than two. MeshFunction will have a particular interface to deal with this case. 

First, I will have to implement a method in PointLocator that forces looping through all elements and does not return the first element satisfying the acceptance criteria.

**Comment by jwpeterson** [comment]
> First, I will have to implement a method in PointLocator that forces looping through all elements and does not return the first element satisfying the acceptance criteria.

I think this would be generally useful in other situations as well.  For example, when locating the element a Dirac point is "in", you sometimes need to know *all* possible elements the point can be considered to be in, so that the simulation can agree on a single element when running in parallel and avoid double contributions.

What is your plan for this?  You could take the initial Elem found by the PointLocator and search in a nearby "patch" of elements (neighbors of neighbors, etc) until you think you've found them all...

**Comment by snschune** [comment]
> What is your plan for this? You could take the initial Elem found by the PointLocator and search in a nearby "patch" of elements (neighbors of neighbors, etc) until you think you've found them all...

That's probably much better than what I am doing. You'll see in my PR in a second. I am using the
`close_to_point(Point p, Real tol)` routine and simply don't return in the overloaded versions of 
`PointLocatorBase::operator()`.

**Comment by roystgnr** [comment]
Searching a patch won't do it; that'll fail in the cases of overlapping domains, multi-dimensional meshes, meshes that self-intersect only at a corner...

**Comment by jwpeterson** [comment]
> overlapping domains, multi-dimensional meshes

What does the PointLocator currently do in such cases??

**Comment by roystgnr** [comment]
It's a crapshoot.  It returns the first active element it finds that contains the requested point and belongs to one of the allowed subdomains, and if that element might not be of the same dimension or on the same manifold as the one you want, it's up to you to figure out what to do about that.

I'd really like to see that fixed, too; and doing so might be as simple as crafting TreeNode::find_elements along the lines of find_element but without exiting as soon as we have a single success.

**Comment by pbauman** [comment]
> It returns the first active element it finds that contains the requested point and belongs to one of the allowed subdomains, and if that element might not be of the same dimension or on the same manifold as the one you want, it's up to you to figure out what to do about that.

I'm having trouble finding the discussion, but IIRC, we'd said that a reasonable assumption is that different manifolds should have a different subdomain ids and, in that case, there should be no ambiguity here as long as the user uses the corresponding subdomain id. I think the same thing can be said of overlapping domains, too.

**Comment by friedmud** [comment]
I agree with Paul.

**Comment by roystgnr** [comment]
Yeah, thanks Paul.  I'd forgotten that conversation, but we did come to the right conclusion: requiring the user to keep their subdomain ids manifold-specific is less of an imposition than cooking up and requiring them to learn some new manifold-selection API.
